src/folding/circuits/sum_check.rs

```diff
@@ -0,0 +1,176 @@
+    use ark_r1cs_std::{
+        alloc::{AllocVar, AllocationMode},
+        fields::fp::FpVar,
+    };
+    use ark_relations::r1cs::ConstraintSystem;
+    use std::sync::Arc;
+
+    use super::SumCheckVerifierGadget;
+
+    pub type TestSumCheckProof<F> = (VirtualPolynomial<F>, PoseidonConfig<F>, IOPProof<F>);
+
+    /// Primarily used for testing the sumcheck gadget
+    /// Returns a random virtual polynomial, the poseidon config used and the associated sumcheck proof
+    pub fn get_test_sumcheck_proof<C: CurveGroup>(
+        num_vars: usize,
+    ) -> TestSumCheckProof<C::ScalarField>
+    where
+        <C as ark_ec::Group>::ScalarField: Absorb,
+    {
+        let mut rng = ark_std::test_rng();
+        let poseidon_config: PoseidonConfig<C::ScalarField> =
+            poseidon_test_config::<C::ScalarField>();
+        let mut poseidon_transcript_prove = PoseidonTranscript::<C>::new(&poseidon_config);
+        let poly_mle = DenseMultilinearExtension::rand(num_vars, &mut rng);
+        let virtual_poly =
+            VirtualPolynomial::new_from_mle(&Arc::new(poly_mle), C::ScalarField::ONE);
+        let sum_check: IOPProof<C::ScalarField> = IOPSumCheck::<C, PoseidonTranscript<C>>::prove(
+            &virtual_poly,
+            &mut poseidon_transcript_prove,
+        )
+        .unwrap();
+        (virtual_poly, poseidon_config, sum_check)
+    }
+
+    #[test]
+    fn test_sum_check_circuit() {
+        for num_vars in 1..15 {
+            let cs = ConstraintSystem::<Fr>::new_ref();
+            let (virtual_poly, poseidon_config, sum_check) =
+                get_test_sumcheck_proof::<Projective>(num_vars);
+            // initiate univariate polynomial variables
+            let poly_vars = sum_check.get_poly_vars_from_sumcheck_proof(cs.clone());
+            let poly_num_variables_var = FpVar::new_variable(
+                cs.clone(),
+                || Ok(Fr::from(virtual_poly.aux_info.num_variables as u64)),
+                AllocationMode::Witness,
+            )
+            .unwrap();
+            let poly_max_degree_var = FpVar::new_variable(
+                cs.clone(),
+                || Ok(Fr::from(virtual_poly.aux_info.max_degree as u64)),
+                AllocationMode::Witness,
+            )
+            .unwrap();
+            let mut poseidon_var = PoseidonTranscriptVar::new(cs.clone(), &poseidon_config);
+            poseidon_var.absorb(poly_num_variables_var).unwrap();
+            poseidon_var.absorb(poly_max_degree_var).unwrap();
+            let claim =
+                IOPSumCheck::<Projective, PoseidonTranscript<Projective>>::extract_sum(&sum_check);
+            let claim_var =
+                FpVar::new_variable(cs.clone(), || Ok(claim), AllocationMode::Witness).unwrap();
+            let res = SumCheckVerifierGadget::verify(&poly_vars, &claim_var, &mut poseidon_var);
+            assert!(res.is_ok());
+            assert!(cs.is_satisfied().unwrap());
```

Comment:
Should we check that the computed & returned values from `verify` are correct? (the `e_var, r_vars`)

Reply:
Yup, implemented additional checks [here](https://github.com/privacy-scaling-explorations/folding-schemes/pull/47/commits/209e9c800f0d469a862cc580924c653c73623214)

Reply:
Nice!! ^^